### PR TITLE
Feature : Two ways to use library

### PR DESCRIPTION
**Pull request by bharathyes**
* Allow library use by extending class or via static method
* Assert test by object ID in result data
* Cucumber scenarios reuse w/ Outline & Examples (datatables)

### CODE REVIEW

File: src/main/java/com/techconative/inmemory/pagination/util/InmemoryFOPS.java [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+import com.techconative.inmemory.pagination.modal.PaginationCriteria;
+import exception.InvalidCriteriaException;
+import java.util.*;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * Filter, search, sort and paginate your data. PaginationService is the class to be extended to
```
Comment (vivek-dhayalan) [inline src/main/java/com/techconative/inmemory/pagination/util/InmemoryFOPS.java:16]: @bharathyes check the comments please and update accordingly.

File: src/main/java/com/techconative/inmemory/pagination/util/InmemoryFOPS.java [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+    public static PageResult processData(List rawData, PaginationCriteria criteria) {
+        List<Map<String, String>> data = convert(rawData);
+
+        if (criteria.getFilter() != null && !criteria.getFilter().isEmpty()) {
+            data = applyFiltering(criteria, data);
+        }
+
+        LinkedList<Map<String, String>> sortedList = applySorting(criteria, data);
+
+        LinkedList resultList = applyPagination(criteria, sortedList);
```
Comment (vivek-dhayalan) [inline src/main/java/com/techconative/inmemory/pagination/util/InmemoryFOPS.java:44]: I'm not looking into the logics inside as I assumed these logics are moved from one class to other.

File: src/test/resources/cucumber/pagination/CounterTest.feature [hunk 3]
```diff
@@ -1,11 +1,15 @@
...
-    When Complex filtering criteria is "counters.[].yearMonth.[].dateCounter.[].likes=10&*=Vega|vegas"
-    Then Result size should be 0
\ No newline at end of file
+  Scenario Outline: Deep Nested filtering test in counter data
+    When Complex filtering criteria is <criteria>
+    Then Result size should be <expectedSize>
+    And <objId> present in <expectedObjs>
+
+    Examples:
+      | criteria | expectedSize | objId | expectedObjs |
```
Comment (vivek-dhayalan) [inline src/test/resources/cucumber/pagination/CounterTest.feature:13]: is this objId or key to be inspected?

File: src/main/java/com/techconative/inmemory/pagination/util/InmemoryFOPS.java (hunk 1 above)
Comment (bharathyes) [reply src/main/java/com/techconative/inmemory/pagination/util/InmemoryFOPS.java]: Fixed the comments 👍🏾

File: src/main/java/com/techconative/inmemory/pagination/util/InmemoryFOPS.java (hunk 2 above)
Comment (bharathyes) [reply src/main/java/com/techconative/inmemory/pagination/util/InmemoryFOPS.java:42]: Yes. Beyond a few minor edits most are just move to the new class.

File: src/main/java/com/techconative/inmemory/pagination/util/InmemoryFOPS.java [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+            PaginationCriteria criteria, List<Map<String, String>> filteredList) {
+        LinkedList<Map<String, String>> sortedList = null;
+        if (criteria.getSort().equals(OrderingCriteria.DESC)) {
+            sortedList =
+                    filteredList.stream()
+                            .sorted(
+                                    Comparator.comparing(
+                                            m -> String.valueOf(m.get(criteria.getColumn())),
+                                            Comparator.nullsFirst(Comparator.reverseOrder())))
+                            .collect(Collectors.toCollection(LinkedList::new));
```
[the remark this answers is not in the document]
Comment (bharathyes) [reply src/main/java/com/techconative/inmemory/pagination/util/InmemoryFOPS.java:219]: @sonatype-lift ignoreall

File: src/test/resources/cucumber/pagination/CounterTest.feature (hunk 3 above)
Comment (vivek-dhayalan) [reply src/test/resources/cucumber/pagination/CounterTest.feature:13]: yes if possible

Comment (bharathyes) [reply src/test/resources/cucumber/pagination/CounterTest.feature]: Changed this step to `And Object identifier <objID> is present in <expectedObjsList>`

**Review by vivek-dhayalan — APPROVED** [review]

**Pull request merged**